File: Dart/permutations.dart
class Solution {
  List<List<int>> permute(List<int> nums) {
    List<List<int>> result = [];
    int length = nums.length;
    void backtrack(int start) {
      if (start == length) {
        result.add(List.from(nums));
      }
      for (int i = start; i < length; i++) {
        swap(nums, start, i);
        backtrack(start + 1);
        swap(nums, start, i);
      }
    }

    backtrack(0);
    return result;
  }

  void swap(List<int> nums, int a, int b) {
    int temp = nums[a];
    nums[a] = nums[b];
    nums[b] = temp;
  }
}

void main() {
  Solution solution = Solution();
  List<int> nums = [1, 2, 3];
  List<List<int>> permutations = solution.permute(nums);
  print(permutations);
}

File: Python/Permutations.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        result: List[List[int]] = []
        len_nums: int = len(nums)

        def backtrack(start: int):
            nonlocal result, len_nums
            if start == len_nums:
                result.append(nums[:])
                return
            for i in range(start, len_nums):
                nums[start], nums[i] = nums[i], nums[start]
                backtrack(start + 1)
                nums[start], nums[i] = nums[i], nums[start]

        backtrack(0)
        return result


if __name__ == "__main__":
    s = Solution()
    nums = [1, 2, 3]
    print(s.permute(nums))
